UnitystationLauncher/ContentScanning/AssemblyTypeChecker.cs

```diff
@@ -0,0 +1,1211 @@
+                                var sig = memRef.DecodeMethodSignature(new TypeProvider(), 0);
+
+                                memberRef = new ScanningTypes.MMemberRefMethod(
+                                    parent,
+                                    memName,
+                                    sig.ReturnType,
+                                    sig.GenericParameterCount,
+                                    sig.ParameterTypes);
+
+                                break;
+                            }
+                        case MemberReferenceKind.Field:
+                            {
+                                var fieldType = memRef.DecodeFieldSignature(new TypeProvider(), 0);
+                                memberRef = new ScanningTypes.MMemberRefField(parent, memName, fieldType);
+                                break;
+                            }
+                        default:
+                            throw new ArgumentOutOfRangeException();
+                    }
+
+                    return memberRef;
+                })
+                .Where(p => p != null)
+                .ToList()!;
+        }
+    }
+
+    private List<(MType type, MType parent, ArraySegment<MType> interfaceImpls)> GetExternalInheritedTypes(
+        MetadataReader reader,
+        ConcurrentBag<SandboxError> errors)
+    {
+        var list = new List<(MType, MType, ArraySegment<MType>)>();
+        foreach (var typeDefHandle in reader.TypeDefinitions)
+        {
+            var typeDef = reader.GetTypeDefinition(typeDefHandle);
+            ArraySegment<MType> interfaceImpls;
+            ScanningTypes.MTypeDefined type = GetTypeFromDefinition(reader, typeDefHandle);
+
+            if (!ParseInheritType(type, typeDef.BaseType, out var parent, reader, errors))
+            {
+                continue;
+            }
+
+            var interfaceImplsCollection = typeDef.GetInterfaceImplementations();
+            if (interfaceImplsCollection.Count == 0)
+            {
+                interfaceImpls = Array.Empty<MType>();
+            }
+            else
+            {
+                interfaceImpls = new MType[interfaceImplsCollection.Count];
+                var i = 0;
+                foreach (var implHandle in interfaceImplsCollection)
+                {
+                    var interfaceImpl = reader.GetInterfaceImplementation(implHandle);
+
+                    if (ParseInheritType(type, interfaceImpl.Interface, out var implemented, reader, errors))
+                    {
+                        interfaceImpls[i++] = implemented;
+                    }
+                }
+
+                interfaceImpls = interfaceImpls[..i];
+            }
+
+            list.Add((type, parent, interfaceImpls));
+        }
+
+        return list;
+    }
+
+    private bool ParseInheritType(MType ownerType, EntityHandle handle, [NotNullWhen(true)] out MType? type, MetadataReader reader, ConcurrentBag<SandboxError> errors)
```

Comment:
Is this `[NotNullWhen(true)]` just to get rid of a warning? I'm not sure why its here

Reply:
looks like it's inherited from SS14 code